db/bigtable_common.go

```diff
@@ -0,0 +1,133 @@
+package db
+
+import (
+	"context"
+	"eth2-exporter/types"
+	"eth2-exporter/utils"
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	gcp_bigtable "cloud.google.com/go/bigtable"
+)
+
+func (bigtable *Bigtable) WriteBulk(mutations *types.BulkMutations, table *gcp_bigtable.Table, batchSize int) error {
+
+	callingFunctionName := utils.GetParentFuncName()
+
+	ctx, done := context.WithTimeout(context.Background(), time.Minute*5)
+	defer done()
+
+	// pre-sort mutations for efficient bulk inserts
+	sort.Sort(mutations)
+	length := batchSize
+	numMutations := len(mutations.Muts)
+	numKeys := len(mutations.Keys)
+	iterations := numKeys / length
+
+	if numKeys != numMutations {
+		return fmt.Errorf("error expected same number of keys as mutations keys: %v mutations: %v", numKeys, numMutations)
+	}
+
+	for offset := 0; offset < iterations; offset++ {
+		start := offset * length
+		end := offset*length + length
+
+		startTime := time.Now()
+		errs, err := table.ApplyBulk(ctx, mutations.Keys[start:end], mutations.Muts[start:end])
+		for _, e := range errs {
+			if e != nil {
+				return err
+			}
+		}
+		logger.Infof("%s: wrote from %v to %v rows to bigtable in %.1f s", callingFunctionName, start, end, time.Since(startTime).Seconds())
+		if err != nil {
+			return err
+		}
+	}
+
+	if (iterations * length) < numKeys {
+		start := iterations * length
+		startTime := time.Now()
+		errs, err := table.ApplyBulk(ctx, mutations.Keys[start:], mutations.Muts[start:])
+		if err != nil {
+			return err
+		}
+		for _, e := range errs {
+			if e != nil {
+				return e
+			}
+		}
+		logger.Infof("%s: wrote from %v to %v rows to bigtable in %.1fs", callingFunctionName, start, numKeys, time.Since(startTime).Seconds())
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	return nil
+}
+
+func (bigtable *Bigtable) DeleteRowsWithPrefix(prefix string) {
```

Comment:
This is currently not used and does require some refactoring. I suggest to delete it. Once it is used, it can be implemented properly with your new `WriteBulk` functionality.

Reply:
deleted in favor of the ClearByPrefix functions which does the same